Build IPA command

jmagman opened the pull request:
## Description

Add a parameter to `flutter build ipa` to pass to `xcodebuild -exportArchive` to generate an IPA from an xcarchive.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/13065
https://github.com/flutter/flutter/pull/67598
Follow up documentation at https://github.com/flutter/website/issues/4852

## Tests

build_ipa_test

## Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change
- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

jmagman committed:
Build IPA command

jonahwilliams commented:
Requiring users to run `flutter build xcarchive` and then `flutter build ipa` could enable some mistakes if users aren't careful. i.e. If the archive is not up to date, could users accidentally upload a stale build?

Similarly, if all CI build scripts would need to be something like `flutter build xcarchive && flutter build ipa`, then perhaps `flutter build ipa` would imply the archiving?

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/build_ipa_test.dart:
```diff
@@ -0,0 +1,323 @@
+      processManager: FakeProcessManager.any(),
+    );
+
+    await expectToolExitLater(
```
Til about expectToolExitLater

jmagman commented:
> Requiring users to run `flutter build xcarchive` and then `flutter build ipa` could enable some mistakes if users aren't careful. i.e. If the archive is not up to date, could users accidentally upload a stale build?

I was taking cues from `install` here, it has similar problems of staleness, it bails if the app binary is missing, and prompts the user to run another `flutter` command:
https://github.com/flutter/flutter/blob/345188d40e576ffb6c97fee60446af58b8c9965a/packages/flutter_tools/lib/src/ios/devices.dart#L255

I could update this to run `xcarchive` if the archive isn't in the expected spot, though it doesn't seem that big a deal to prompt the user to run the right command (with all the flavor, dart, etc flags they need).

I was thinking of this command is a [thin convenience wrapper](https://github.com/flutter/flutter/issues/13065#issuecomment-698667902) on top of an `xcodebuild` command now that `flutter build xcarchive` exists.

jonahwilliams commented:
install probably isn't the best designed `flutter` command though

jonahwilliams commented:
One way you could support this as either the primary development command or a wrapper is allow the user to provide a --path-to-xcarchive (is this customizable) and then use it as is if provided.

xster commented on packages/flutter_tools/lib/src/application_package.dart:
```diff
@@ -385,6 +385,11 @@ class BuildableIOSApp extends IOSApp {
   String get archiveBundlePath
     => globals.fs.path.join(getIosBuildDirectory(), 'archive', globals.fs.path.withoutExtension(_hostAppBundleName));
```
slightly tangential: at some point we might want to take stock of all the "temporary but user accessible" build artifacts mixed with build intermediaries in the build/ folder and organize them. 

Right now I have 

```
build
├── 3f44400bd8f8c7820069645ed94a731d
├── app
│   ├── generated
│   │   ├── ap_generated_sources
│   │   ├── not_namespaced_r_class_sources
│   │   ├── res
│   │   └── source
│   ├── intermediates
│   │   ├── annotation_processor_list
│   │   ├── apk_list
│   │   ├── blame
│   │   ├── bundle_manifest
│   │   ├── check_manifest_result
│   │   ├── compatible_screen_manifest
│   │   ├── dex
│   │   ├── duplicate_classes_check
│   │   ├── external_file_lib_dex_archives
│   │   ├── external_libs_dex
│   │   ├── flutter
│   │   ├── generated_proguard_file
│   │   ├── incremental
│   │   ├── instant_app_manifest
│   │   ├── javac
│   │   ├── lint_jar
│   │   ├── manifest_merge_blame_file
│   │   ├── merged-not-compiled-resources
│   │   ├── merged_assets
│   │   ├── merged_java_res
│   │   ├── merged_jni_libs
│   │   ├── merged_manifests
│   │   ├── merged_native_libs
│   │   ├── merged_shaders
│   │   ├── metadata_feature_manifest
│   │   ├── processed_res
│   │   ├── proguard-files
│   │   ├── proguard-rules
│   │   ├── res
│   │   ├── res_stripped
│   │   ├── shader_assets
│   │   ├── signing_config
│   │   ├── stripped_native_libs
│   │   ├── symbols
│   │   ├── transforms
│   │   └── validate_signing_config
│   ├── outputs
│   │   ├── apk
│   │   ├── flutter-apk
│   │   ├── logs
│   │   └── mapping
│   ├── reports
│   └── tmp
│       ├── compileDebugJavaWithJavac
│       ├── compileProfileJavaWithJavac
│       ├── compileReleaseJavaWithJavac
│       ├── packLibsflutterBuildDebug
│       ├── packLibsflutterBuildProfile
│       └── packLibsflutterBuildRelease
├── fa37e11e55641aa14c7c94ec233f11be
└── ios
    ├── Debug-iphoneos
    │   └── Runner.app
    ├── Profile-iphoneos
    │   ├── Runner.app
    │   └── Runner.app.dSYM
    ├── Runner.build
    │   └── Release-iphoneos
    └── iphoneos
        └── Runner.app
```

without counting ios-frameworks, aars etc. We don't have to fix anything here but it's a question for the next comment.

xster commented on packages/flutter_tools/lib/src/application_package.dart:
```diff
@@ -385,6 +385,11 @@ class BuildableIOSApp extends IOSApp {
+  // The output xcarchive bundle path `build/ios/archive/Runner.xcarchive`.
+  String get archiveBundleOutputPath => globals.fs.path.setExtension(archiveBundlePath, '.xcarchive');
+
+  String get ipaOutputPath => globals.fs.path.join(getIosBuildDirectory(), 'ipa');
```
It looks like the dimensions we have are device type and build modes. I suppose it's non-sensical to have debug ipas or simulator ipas. Was that the thinking process? Should user's own scheme/flavor be in there somewhere?

jmagman replied:
That's what the `flutter build xcarchive` is for.  That was kind of my point about this IPA command being [kind of useless](https://github.com/flutter/flutter/issues/13065#issuecomment-698667902).  A lot of the requests in https://github.com/flutter/flutter/issues/13065 are solved by `flutter build xcarchive`, which has all the compiled code that encapsulates flavor, scheme, etc.
The build mode here only so the tool can find the right directory for their non-release archive and to find the right archive name.

xster commented on packages/flutter_tools/lib/src/commands/build_ipa.dart:
```diff
@@ -0,0 +1,154 @@
+      'export-options-plist',
+      valueHelp: 'ExportOptions.plist',
+      help:
+          'Path to the "-exportOptionsPlist" file passed to Xcode. See "man xcodebuild" for available keys.',
```
For first timers (which we will have), we might need to link to a Flutter page to describe what it is. Since we'll certainly have Flutter developers shipping for the first time to the app store without having shipped an ios app in the past.

Feel free to come back in a later PR too instead of leaving a placeholder.

jmagman replied:
For sure, we'll need to point to some docs.  But we can pretty much just point to Apple's docs since this command is a really thin wrapper and doesn't process any plist values.

jmagman replied:
Added a TODO to update the new option help text.

xster commented on packages/flutter_tools/lib/src/commands/build_ipa.dart:
```diff
@@ -0,0 +1,154 @@
+
+  @override
+  final String description =
+      'Build an iOS App Store package after running "flutter build xcarchive" (Mac OS X host only).';
```
Makes sense that it's 2 separate steps in xcode and it's safer to model our modularity after the underlaying tool. But on the frontend, we don't have to expose the same granularity to the users I suspect. 

Could we just wrap the inner workings?

jmagman replied:
It sounds like that's what you and @jonahwilliams would prefer.

I could convert this command into something that runs in response to the export plist being passed into the `xcarchive` command?  So no `flutter build ipa` but instead `flutter build xcarchive --exportOptionsPlist <file>` that creates the xcarchive and then makes an ipa out of it?

jmagman replied:
Done.

xster commented on packages/flutter_tools/lib/src/commands/build_ipa.dart:
```diff
@@ -0,0 +1,154 @@
+    final FileSystemEntityType type = _fileSystem.typeSync(exportOptionsPlist);
+    if (type == FileSystemEntityType.notFound) {
+      throwToolExit(
+          '"$exportOptionsPlist" property list does not exist. See "man xcodebuild" for available keys.');
```
I don't actually see the available keys in the manual (which is really long). We might need to do a bit more hand-holding for pure-flutter users.

jmagman replied:
You're right, it's in the `xcrun xcodebuild -h` but not the man page.

xster commented on packages/flutter_tools/lib/src/commands/build_ipa.dart:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    final BuildInfo buildInfo =
+        getBuildInfo(forcedBuildMode: BuildMode.release);
```
Can we show a warning even if we make an implicit change?

xster commented on packages/flutter_tools/lib/src/commands/build_ipa.dart:
```diff
@@ -0,0 +1,154 @@
+      final String flavorFlag =
+          buildInfo.flavor == null ? '' : ' --flavor ${buildInfo.flavor}';
+      throwToolExit('xcarchive not found at $archiveBundleOutputPath. '
+          'Run "flutter build xcarchive$flavorFlag" to generate it.');
```
should there be a space in there?

jmagman replied:
The space is in the flavorFlag  `' --flavor ${buildInfo.flavor}'`

xster commented on packages/flutter_tools/lib/src/commands/build_ipa.dart:
```diff
@@ -0,0 +1,154 @@
+          buildInfo.flavor == null ? '' : ' --flavor ${buildInfo.flavor}';
+      throwToolExit('xcarchive not found at $archiveBundleOutputPath. '
+          'Run "flutter build xcarchive$flavorFlag" to generate it.');
+    }
```
if we do wrap the xcarchive part, it would be nice to have some descriptions around what the flavor translates to in the UI if the asked flavor isn't present.

jmagman commented:
I think I'm being too pedantic differentiating between building xcarchive and ipa.  I am going to combine the commands so it's not a two-step that can get out of sync.

xster commented:
+1. We should replicate Apple's capabilities but we don't have to replicate their exact APIs if there are opportunities to make it clearer.

jmagman committed:
xcarchive -> ipa

jmagman commented:
Reworked this based on feedback.

Renamed `flutter build xcarchive` to `flutter build ipa` (but aliased to `xcarchive` because I like it better and already said it was happening).  Add a `--export-options-plist` option that takes the `xcarchive` that was just built (with all the flavor, dart flags, build info, etc) and additionally creates an IPA out of it, so it's no longer two commands.

I will update the publishing iOS docs when this lands and update the help text to link to something more useful than `xcodebuild -h`.

xster commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -56,14 +57,27 @@ class BuildIOSCommand extends _BuildIOSSubCommand {
 }
 
-/// Builds an .xcarchive for an iOS app to be generated for App Store submission.
+/// Builds an .xcarchive and optionally .ipa for an iOS app to be generated for
```
sounds like it's the other way around now?

jmagman replied:
No, it always builds an .xcarchive, and only builds an IPA if you pass in the plist flag.

xster commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -79,6 +93,76 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
+      final FileSystemEntityType type = globals.fs.typeSync(exportOptionsPlist);
+      if (type == FileSystemEntityType.notFound) {
+        throwToolExit(
+            '"$exportOptionsPlist" property list does not exist. See "man xcodebuild" for available keys.');
```
The second sentence here and below seem like non sequiturs? I assume you're just trying to say file x you specified doesn't exist?

Also be consistent with man xcodebuild vs xcodebuild -h

jmagman replied:
Sorry I saw this after it was merged.  Will follow up.

jmagman replied:
https://github.com/flutter/flutter/pull/68826

xster reviewed (commented):
LGTM

jonahwilliams commented on packages/flutter_tools/lib/src/commands/build_ios.dart:
```diff
@@ -79,6 +93,76 @@ class BuildIOSArchiveCommand extends _BuildIOSSubCommand {
+      // Error Domain=IDEFoundationErrorDomain Code=1 "exportOptionsPlist error for key 'method': expected one of {app-store, ad-hoc, enterprise, development, validation}, but found developmentasdasd" ...
+      LineSplitter.split(result.stderr)
+          .where((String line) => line.contains('error: '))
+          .forEach(errorMessage.writeln);
```
nice use of tear-offs!

jonahwilliams reviewed (approved):
LGTM

shinriyo commented:
Woooow!!! nice feature! THANK YOU!
